src/devices/video/v9938.cpp

```diff
@@ -1138,76 +1138,123 @@ void v99x8_device::mode_graphic1(uint32_t *ln, int line)
+	const pen_t border_pen = pen16(m_cont_reg[7] & 0x0f);
 
-	colourmask = ((m_cont_reg[3] & 0x7f) * 8) | 7;
-	patternmask = ((m_cont_reg[4] & 0x03) * 256) | 0xff;
+	int nametbl_base = (m_cont_reg[2] << 10) + ((scrolled_y / 8) * 32);
+	int nametbl_offset = (m_cont_reg[26] & 0x1f);
 
-	nametbl_addr =  (m_cont_reg[2] << 10);
-	colourtbl_addr =  ((m_cont_reg[3] & 0x80) << 6) + (m_cont_reg[10] << 14);
-	patterntbl_addr = ((m_cont_reg[4] & 0x3c) << 11);
+	if (BIT(m_cont_reg[25], 0) && BIT(m_cont_reg[26], 5))
+		nametbl_base ^= 0x8000;
 
-	line2 = (line + m_cont_reg[23]) & 255;
-	name = (line2/8)*32;
+	for (int x = m_offset_x * 2; x > 0; x--)
+		*ln++ = border_pen;
 
-	pen = pen16(m_cont_reg[7] & 0x0f);
-	xxx = m_offset_x * 2;
-	while (xxx--) *ln++ = pen;
+	int dot_scroll = m_cont_reg[27] & 0x07;
+	int pixels_to_mask = BIT(m_cont_reg[25], 1) ? 8 - dot_scroll : 0;
+	int pixels_to_draw = 256 - dot_scroll;
+	while (dot_scroll--)
+	{
+		*ln++ = border_pen;
+		*ln++ = border_pen;
+	}
 
-	for (x=0;x<32;x++)
+	do
 	{
-		charcode = m_vram_space->read_byte(nametbl_addr + name) + (line2&0xc0)*4;
-		colour = m_vram_space->read_byte(colourtbl_addr + ((charcode&colourmask)*8+(line2&7)));
-		pattern = m_vram_space->read_byte(patterntbl_addr + ((charcode&patternmask)*8+(line2&7)));
-		fg = pen16(colour >> 4);
-		bg = pen16(colour & 15);
-		for (xx=0;xx<8;xx++)
+		const int charcode = m_vram_space->read_byte(nametbl_base + (nametbl_offset & 0x1f)) + (scrolled_y & 0xc0) * 4;
+		const u8 colour = m_vram_space->read_byte(colourtbl_addr + ((charcode & colourmask) * 8 + (scrolled_y & 7)));
+		u8 pattern = m_vram_space->read_byte(patterntbl_addr + ((charcode & patternmask) * 8 + (scrolled_y & 7)));
+		const pen_t fg = pen16(colour >> 4);
+		const pen_t bg = pen16(colour & 15);
+		for (int x = 0; x < 8 && pixels_to_draw > 0; x++)
 		{
-			*ln++ = (pattern & 0x80) ? fg : bg;
-			*ln++ = (pattern & 0x80) ? fg : bg;
+			if (!pixels_to_mask)
+			{
+				*ln++ = (pattern & 0x80) ? fg : bg;
+				*ln++ = (pattern & 0x80) ? fg : bg;
+			}
+			else
+			{
+				pixels_to_mask--;
+				*ln++ = border_pen;
+				*ln++ = border_pen;
+			}
+			pixels_to_draw--;
 			pattern <<= 1;
 		}
-		name++;
+		nametbl_offset = (nametbl_offset + 1) & 0x1f;
+		if (BIT(m_cont_reg[25], 0) && !nametbl_offset)
+			nametbl_base ^= 0x8000;
 	}
+	while (pixels_to_draw > 0);
 
-	xx = (16 - m_offset_x) * 2;
-	while (xx--) *ln++ = pen;
+	for (int x = (16 - m_offset_x) * 2; x > 0; x--)
+		*ln++ = border_pen;
 }
 
 void v99x8_device::mode_graphic4(uint32_t *ln, int line)
 {
-	int nametbl_addr, colour;
-	int line2, linemask, x, xx;
-	pen_t pen, pen_bg;
+	int linemask = ((m_cont_reg[2] & 0x1f) << 3) | 7;
+	const int scrolled_y = ((line + m_cont_reg[23]) & linemask) & 255;
+	const pen_t border_pen = pen16(m_cont_reg[7] & 0x0f);
 
-	linemask = ((m_cont_reg[2] & 0x1f) << 3) | 7;
+	int nametbl_base = ((m_cont_reg[2] & 0x40) << 10) + scrolled_y * 128;
+	int nametbl_offset = (m_cont_reg[26] & 0x1f) * 4;
```

Comment:
This would be clearer with shifts, too.

Reply:
Changed all occurrences to shifts and hexadecimal masks